File: sms/src/main/java/dev/pitlor/sms/repositories/MessageRepository.java
package dev.pitlor.sms.repositories;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import dev.pitlor.sms.models.Mms;
import dev.pitlor.sms.models.Sms;

public class MessageRepository {
    private static final Uri CONVERSATIONS_URI = Uri.parse("content://mms-sms/conversations/");
    public static final int SMS = 1;
    public static final int MMS = 2;

    private Context context;

    @Inject
    public MessageRepository(Context context) {
        this.context = context;
    }

    public List<List<String>> getAllIdsAfter(OffsetDateTime minimumTime) {
        ContentResolver contentResolver = context.getContentResolver();
        final String[] projection = new String[]{"_id", "ct_t"};

        List<String> smsIds = new ArrayList<>();
        List<String> mmsIds = new ArrayList<>();
        Cursor cursor = contentResolver.query(CONVERSATIONS_URI, projection, null, null, null);
        if (cursor != null && cursor.moveToFirst()) {
            int idColumnIndex = cursor.getColumnIndex("_id");
            int typeIndex = cursor.getColumnIndex("ct_t");

            do {
                String string = cursor.getString(typeIndex);
                if ("application/vnd.wap.multipart.related".equals(string)) {
                    mmsIds.add(cursor.getString(idColumnIndex));
                } else {
                    smsIds.add(cursor.getString(idColumnIndex));
                }
            } while (cursor.moveToNext());

            cursor.close();
        }

        List<List<String>> result = new ArrayList<>();
        result.add(smsIds);
        result.add(mmsIds);
        return result;
    }

    public Sms getSmsById(String id) {
//        Sms sms = Sms.Companion.Builder();
//        ContentResolver contentResolver = context.getContentResolver();
//        Cursor cursor = contentResolver.query(
//            Telephony.Sms.CONTENT_URI,
//            new String[]{THREAD_ID, ADDRESS, BODY, DATE, SUBJECT},
//            Telephony.Sms._ID + " = ?",
//            new String[]{id},
//            null
//        );
//        if (cursor != null && cursor.moveToFirst()) {
//            sms
//                .address(cursor.getString(cursor.getColumnIndex(ADDRESS)))
//                .body(cursor.getString(cursor.getColumnIndex(BODY)))
//                .dateReceived(OffsetDateTime.ofInstant(
//                    Instant.ofEpochMilli(cursor.getLong(cursor.getColumnIndex(DATE))), ZoneId.systemDefault()
//                ))
//                .threadId(cursor.getLong(cursor.getColumnIndex(THREAD_ID)))
//                .subject(cursor.getString(cursor.getColumnIndex(SUBJECT)));
//            cursor.close();
//        }
//
//        return sms.build();
        return null;
    }

    public Mms getMmsById(String id) {
//        Mms.MmsBuilder mms = Mms.builder();
//        ContentResolver contentResolver = context.getContentResolver();
//        Cursor cursor;
//
//        // Metadata
//        cursor = contentResolver.query(
//            Telephony.Mms.CONTENT_URI,
//            new String[]{Telephony.Mms.DATE, Telephony.Mms.SUBJECT, Telephony.Mms.THREAD_ID},
//            "_id = ?",
//            new String[]{id},
//            null
//        );
//        if (cursor != null && cursor.moveToFirst()) {
//            String subject = cursor.getString(cursor.getColumnIndex(Telephony.Mms.SUBJECT));
//            Instant dateRecieved = Instant.ofEpochSecond(cursor.getLong(cursor.getColumnIndex(Telephony.Mms.DATE)));
//
//            mms
//                .dateReceived(OffsetDateTime.ofInstant(dateRecieved, ZoneId.systemDefault()))
//                .subject(subject != null ? subject : "")
//                .threadId(cursor.getLong(cursor.getColumnIndex(Telephony.Mms.THREAD_ID)));
//
//            cursor.close();
//        }
//
//        // Text/Picture
//        cursor = contentResolver.query(
//            Telephony.Mms.Part.CONTENT_URI,
//            null,
//            Telephony.Mms.Part.MSG_ID + " = ?",
//            new String[]{id},
//            null
//        );
//        if (cursor != null && cursor.moveToFirst()) {
//            StringBuilder bodyBuilder = new StringBuilder();
//            do {
//                switch (cursor.getString(cursor.getColumnIndex(Telephony.Mms.Part.CONTENT_TYPE))) {
//                    case "text/plain":
//                        String data = cursor.getString(cursor.getColumnIndex(Telephony.Mms.Part._DATA));
//                        String body;
//                        if (data != null) {
//                            body = getText(context, cursor.getString(cursor.getColumnIndex(Telephony.Mms.Part._ID)));
//                        } else {
//                            body = cursor.getString(cursor.getColumnIndex(Telephony.Mms.Part.TEXT));
//                        }
//                        bodyBuilder.append(body);
//                        mms.body(bodyBuilder.toString());
//                        break;
//                    case "image/jpeg":
//                    case "image/bmp":
//                    case "image/gif":
//                    case "image/jpg":
//                    case "image/png":
//                        File picture = getImage(context, cursor.getString(cursor.getColumnIndex(Telephony.Mms._ID)), id);
//                        mms.picture(picture);
//                        break;
//                }
//            } while (cursor.moveToNext());
//            cursor.close();
//        }
//
//        // Sender
//        cursor = contentResolver.query(
//            Uri.parse(MessageFormat.format("content://mms/{0}/addr", id)),
//            null,
//            Telephony.Mms.Addr.MSG_ID + " = ?",
//            new String[]{id},
//            null
//        );
//        if (cursor != null && cursor.moveToFirst()) {
//            do {
//                String number = cursor.getString(cursor.getColumnIndex(Telephony.Mms.Addr.ADDRESS));
//                if (number != null) {
//                    mms.address(number);
//                    break;
//                }
//            } while (cursor.moveToNext());
//            cursor.close();
//        }
//
//        return mms.build();
        return null;
    }

    private String getText(Context context, String id) {
        InputStream inputStream = null;
        StringBuilder stringBuilder = new StringBuilder();
        try {
            inputStream = context.getContentResolver().openInputStream(Uri.parse("content://mms/part/" + id));
            if (inputStream != null) {
                InputStreamReader isr = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
                BufferedReader reader = new BufferedReader(isr);
                String line = reader.readLine();
                while (line != null) {
                    stringBuilder.append(line);
                    line = reader.readLine();
                }
            }
        } catch (IOException e) {
            // Do nothing
        } finally {
            if (inputStream != null) {
                try {
                    inputStream.close();
                } catch (IOException e) {
                    // Do nothing
                }
            }
        }
        return stringBuilder.toString();
    }

    private File getImage(Context context, String partId, String messageId) {
        File file = new File(context.getCacheDir(), messageId + "_" + partId + ".png");
        if (file.exists()) {
            return file;
        }

        ContentResolver contentResolver = context.getContentResolver();
        try (InputStream inputStream = contentResolver.openInputStream(Uri.parse("content://mms/part/" + partId))) {
            if (inputStream == null) {
                return null;
            }

            byte[] buffer = new byte[inputStream.available()];
            //noinspection ResultOfMethodCallIgnored
            inputStream.read(buffer);
            FileOutputStream outputStream = new FileOutputStream(file);
            outputStream.write(buffer);
        } catch (Exception e) {
            // Do nothing
        }

        return file;
    }
}

File: sms/src/main/java/dev/pitlor/sms/repositories/MessageRepository.kt
package dev.pitlor.sms.repositories

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.provider.Telephony
import dagger.hilt.android.qualifiers.ApplicationContext
import dev.pitlor.sms.*
import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.util.*
import javax.inject.Inject

class MessageRepository @Inject constructor(@ApplicationContext private val context: Context) {
    private val contentResolver = context.contentResolver!!

    suspend fun applyAllMessages(saveMessages: suspend (List<MessageDTO>) -> Unit) {
        val messages = ArrayList<MessageDTO>()
        contentResolver.queryLoop(
            Telephony.MmsSms.CONTENT_CONVERSATIONS_URI,
            projection = arrayOf(Telephony.TextBasedSmsColumns.THREAD_ID),
            isAsync = true
        ) {
            val conversationId = getLong(Telephony.TextBasedSmsColumns.THREAD_ID)
            contentResolver.queryLoop(
                Uri.parse("content://mms-sms/conversations/$conversationId"),
                projection = arrayOf("_id", "ct_t"),
                isAsync = true
            ) {
                val messageDto = MessageDTO(getString("_id"), getString("ct_t") == MMS_TYPE)
                messages.add(messageDto)

                if (messages.size == 100) {
                    saveMessages(messages)
                    messages.clear()
                }
            }
        }
    }

    fun getAllIdsAfter(minimumTime: OffsetDateTime): MessagesDTO {
        val conversations = ArrayList<Long>()
        contentResolver.queryLoop(
            Telephony.MmsSms.CONTENT_CONVERSATIONS_URI,
            projection = arrayOf(Telephony.TextBasedSmsColumns.THREAD_ID),
        ) {
            conversations.add(getLong(Telephony.TextBasedSmsColumns.THREAD_ID))
        }

        val smsIds = ArrayList<String>()
        val mmsIds = ArrayList<String>()
        conversations.forEach { conversationId ->
            contentResolver.queryLoop(
                Uri.parse("content://mms-sms/conversations/$conversationId"),
                projection = arrayOf("_id", "ct_t"),
                selection = "${Telephony.TextBasedSmsColumns.DATE} > ?",
                selectionArgs = arrayOf(minimumTime.toString())
            ) {
                when (getString("ct_t")) {
                    "application/vnd.wap.multipart.related" -> mmsIds.add(getString("_id"))
                    else -> smsIds.add(getString("_id"))
                }
            }
        }

        return MessagesDTO(smsIds, mmsIds)
    }

    fun getSmsById(id: String): Sms {
        return Sms().apply {
            contentResolver.queryOnce(
                Telephony.Sms.CONTENT_URI,
                projection = arrayOf(
                    Telephony.TextBasedSmsColumns.THREAD_ID,
                    Telephony.TextBasedSmsColumns.ADDRESS,
                    Telephony.TextBasedSmsColumns.BODY,
                    Telephony.TextBasedSmsColumns.DATE,
                    Telephony.TextBasedSmsColumns.SUBJECT
                ),
                selection = "${Telephony.Sms._ID} = ?",
                selectionArgs = arrayOf(id)
            ) {
                address = getString(Telephony.TextBasedSmsColumns.ADDRESS)
                body = getString(Telephony.TextBasedSmsColumns.BODY)
                dateReceived = longToDate(getLong(Telephony.TextBasedSmsColumns.DATE))
                threadId = getLong(Telephony.TextBasedSmsColumns.THREAD_ID)
                subject = getString(Telephony.TextBasedSmsColumns.SUBJECT)
            }
        }
    }

    fun getMmsById(id: String): Mms {
        return Mms().apply {
            // Metadata
            contentResolver.queryOnce(
                Telephony.Mms.CONTENT_URI,
                projection = arrayOf(Telephony.Mms.DATE, Telephony.Mms.SUBJECT, Telephony.Mms.THREAD_ID),
                selection = "${Telephony.Mms._ID} = ?",
                selectionArgs = arrayOf(id)
            ) {
                subject = getString(Telephony.Mms.SUBJECT)
                dateReceived = OffsetDateTime.ofInstant(
                    Instant.ofEpochSecond(getLong(Telephony.Mms.DATE)),
                    ZoneId.systemDefault()
                )
                threadId = getLong(Telephony.Mms.THREAD_ID)
            }

            // Text/Picture
            contentResolver.queryLoop(
                Telephony.Mms.Part.CONTENT_URI,
                selection = "${Telephony.Mms.Part.MSG_ID} = ?",
                selectionArgs = arrayOf(id),
            ) {
                when (getString(Telephony.Mms.Part.CONTENT_TYPE)) {
                    "text/plain" -> body += getText(getString(Telephony.Mms.Part._ID))
                    "image/jpeg",
                    "image/bmp",
                    "image/gif",
                    "image/jpg",
                    "image/png" -> picture = getImage(getString(Telephony.Mms._ID))
                }
            }

            // Sender
            contentResolver.queryLoop(
                Uri.parse("content://mms/${id}/addr"),
                selection = "${Telephony.Mms.Addr.MSG_ID} = ?",
                selectionArgs = arrayOf(id),
            ) {
                when (val retrievedAddress = getString(Telephony.Mms.Addr.ADDRESS)) {
                    "" -> noop()
                    else -> address = retrievedAddress
                }
            }
        }
    }

    private fun getText(id: String): String {
        val stringBuilder = StringBuilder()
        contentResolver.useInputStream(Uri.parse("content://mms/part/$id")) {
            val reader = BufferedReader(InputStreamReader(this, StandardCharsets.UTF_8))
            var line = reader.readLine()
            while (line != null) {
                stringBuilder.append(line)
                line = reader.readLine()
            }
        }

        return stringBuilder.toString()
    }

    private fun getImage(partId: String): Bitmap? {
        var bitmap: Bitmap? = null
        contentResolver.useInputStream(Uri.parse("content://mms/part/$partId")) {
            val buffer = ByteArray(available())
            read(buffer)
            bitmap = BitmapFactory.decodeByteArray(buffer, 0, available())
        }

        return bitmap
    }

    companion object {
        const val MMS_TYPE = "application/vnd.wap.multipart.related"
    }
}